pkg/didcomm/common/service/connection_record.go

```diff
@@ -0,0 +1,205 @@
+}
+
+// ConnectionRecord contain info about did exchange connection
+type ConnectionRecord struct {
+	ConnectionID    string
+	State           string
+	ThreadID        string
+	TheirLabel      string
+	TheirDID        string
+	MyDID           string
+	ServiceEndPoint string
+	RecipientKeys   []string
+	InvitationID    string
+	InvitationDID   string
+	Implicit        bool
+	Namespace       string
+}
+
+// NewConnectionRecorder returns new connection recorder instance
+func NewConnectionRecorder(p provider) (*ConnectionRecorder, error) {
+	store, err := p.StorageProvider().OpenStore(nameSpace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new connection recorder : %w", err)
+	}
+
+	transientStore, err := p.TransientStorageProvider().OpenStore(nameSpace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new connection recorder : %w", err)
+	}
+
+	return &ConnectionRecorder{transientStore: transientStore, store: store}, nil
+}
+
+// ConnectionRecorder takes care of connection related persistence features
+type ConnectionRecorder struct {
+	transientStore storage.Store
+	store          storage.Store
+}
+
+// GetConnectionRecord return connection record based on the connection ID
+func (c *ConnectionRecorder) GetConnectionRecord(connectionID string) (*ConnectionRecord, error) {
+	rec, err := getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.store)
+	if err != nil {
+		if errors.Is(err, storage.ErrDataNotFound) {
+			return getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.transientStore)
+		}
+
+		return nil, err
+	}
+
+	return rec, nil
+}
+
+// QueryConnectionRecords returns connection records found in underlying store
+// for given query criteria
+func (c *ConnectionRecorder) QueryConnectionRecords() ([]*ConnectionRecord, error) {
+	// TODO https://github.com/hyperledger/aries-framework-go/issues/655 query criteria to be added as part of issue
+	searchKey := GetConnectionKeyPrefix()("")
+
+	itr := c.store.Iterator(searchKey, fmt.Sprintf(limitPattern, searchKey))
+	defer itr.Release()
+
+	var records []*ConnectionRecord
+
+	keys := make(map[string]struct{})
+
+	for itr.Next() {
+		var record ConnectionRecord
+
+		err := json.Unmarshal(itr.Value(), &record)
+		if err != nil {
+			return nil, fmt.Errorf("failed to query connection records, %w", err)
+		}
+
+		keys[string(itr.Key())] = struct{}{}
+
+		records = append(records, &record)
+	}
+
+	transientItr := c.transientStore.Iterator(searchKey, fmt.Sprintf(limitPattern, searchKey))
+	defer transientItr.Release()
+
+	for transientItr.Next() {
+		// don't fetch data from transient store if same record is present in permanent store
+		if _, ok := keys[string(transientItr.Key())]; !ok {
+			var record ConnectionRecord
+
+			err := json.Unmarshal(transientItr.Value(), &record)
+			if err != nil {
+				return nil, fmt.Errorf("query connection records from transient store : %w", err)
+			}
+
+			records = append(records, &record)
+		}
```

Comment:
I know it is not your code, but better to avoid nested if statements e.g looks better to me

```
if _, ok := keys[string(transientItr.Key())]; ok {
    continue
}

var record ConnectionRecord

if err := json.Unmarshal(transientItr.Value(), &record); err != nil {
    return nil, fmt.Errorf("query connection records from transient store : %w", err)
}

records = append(records, &record)
```

Reply:
Just moved existing code, will fix this.

Reply:
done